Make FX Graph Cache work with distributed training

oulgen opened the pull request:
During distributed training if all ranks except one hit the cache, the rank that did not hit the cache will cause a NCCL timeout since rest of the ranks will enter the collective and start the timer. This PR uses the new PTD API to increase timeout for the ranks that hit the cache by the amount of time the cache would save.


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #133504
* __->__ #133374



cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

Differential Revision: [D61363722](https://our.internmc.facebook.com/intern/diff/D61363722)

oulgen committed:
Make FX Graph Cache work with distributed training

[ghstack-poisoned]

oulgen commented on torch/_inductor/codecache.py:
```diff
@@ -1304,6 +1304,13 @@ def load(  # type: ignore[no-untyped-def]
+                    and (time_taken_ns := compiled_graph._time_taken_ns) is not None
+                ):
+                    dist.distributed_c10d._add_ephemeral_timeout_for_all_pgs(
+                        timedelta(seconds=time_taken_ns // 1e9)  # Convert to seconds
```
On a follow up pr, I’m going to add a fudge factor here to account for differences like 15%

ezyang replied:
Logging here would be good, I think INFO is good to start

c00w commented on torch/_inductor/codecache.py:
```diff
@@ -1304,6 +1304,13 @@ def load(  # type: ignore[no-untyped-def]
                 cache_event_time = time_ns()
+                if (
+                    dist.distributed_c10d.is_initialized()
+                    and (time_taken_ns := compiled_graph._time_taken_ns) is not None
```
Am I reading this right, that this will just be enabled if we land this PR? (Should we put a env var or killswitch on it? I asume we'll want to smoke test it witn an internal job?).

oulgen replied:
Caching is not being used with distributed right now, so this code will not be used anywhere, which is why I dont think we need a killswitch

ezyang commented on torch/_inductor/compile_fx.py:
```diff
@@ -701,6 +701,11 @@ def fx_codegen_and_compile(
+    if (sleep_sec := config.sleep_sec_TESTING_ONLY) is not None:
+        import time
+
+        time.sleep(sleep_sec)
```
Can we log a warning here that the sleep has happened

oulgen committed:
Update on "Make FX Graph Cache work with distributed training"

During distributed training if all ranks except one hit the cache, the rank that did not hit the cache will cause a NCCL timeout since rest of the ranks will enter the collective and start the timer. This PR uses the new PTD API to increase timeout for the ranks that hit the cache by the amount of time the cache would save.





cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

oulgen commented:
@pytorchbot merge

albanD commented:
@pytorchbot revert -m "Broke trunk" -c nosignal

See https://hud.pytorch.org/pytorch/pytorch/commit/dcdb25453e0ddc6a83e0052fffc544d4d03cdffd for example failures.
They were hidden by broken trunk in the last CI run on this PR unfortunately.

oulgen committed:
Update on "Make FX Graph Cache work with distributed training"

During distributed training if all ranks except one hit the cache, the rank that did not hit the cache will cause a NCCL timeout since rest of the ranks will enter the collective and start the timer. This PR uses the new PTD API to increase timeout for the ranks that hit the cache by the amount of time the cache would save.





cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

oulgen committed:
Update on "Make FX Graph Cache work with distributed training"

During distributed training if all ranks except one hit the cache, the rank that did not hit the cache will cause a NCCL timeout since rest of the ranks will enter the collective and start the timer. This PR uses the new PTD API to increase timeout for the ranks that hit the cache by the amount of time the cache would save.





cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

oulgen committed:
Update on "Make FX Graph Cache work with distributed training"

During distributed training if all ranks except one hit the cache, the rank that did not hit the cache will cause a NCCL timeout since rest of the ranks will enter the collective and start the timer. This PR uses the new PTD API to increase timeout for the ranks that hit the cache by the amount of time the cache would save.





cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

oulgen commented:
@oulgen has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D61363722).

oulgen commented:
@oulgen has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D61363722).

oulgen commented:
@pytorchbot merge

eellison commented on torch/_inductor/compile_fx.py:
```diff
@@ -705,6 +705,12 @@ def fx_codegen_and_compile(
     layout_opt: Optional[bool] = None,
     extern_node_serializer: Optional[Callable[[List[ExternKernelNode]], Any]] = None,
 ) -> Union[CompiledFxGraph, str]:
+    if (sleep_sec := config.sleep_sec_TESTING_ONLY) is not None:
+        import time
+
+        log.warning("Sleeping for %s since sleep_sec_TESTING_ONLY is set", sleep_sec)
+        time.sleep(sleep_sec)
```
You could also register a custom operator in test that does the sleeping. I would prefer test code doesnt leaks into source for important functions like this.
